include/range/v3/view/drop_last.hpp

```diff
@@ -0,0 +1,237 @@
+/// \file
+// Range v3 library
+//
+//  Copyright Andrey Diduh 2019
+//
+//  Use, modification and distribution is subject to the
+//  Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/ericniebler/range-v3
+//
+
+#ifndef RANGES_V3_VIEW_DROP_LAST_HPP
+#define RANGES_V3_VIEW_DROP_LAST_HPP
+
+#include <type_traits>
+#include <meta/meta.hpp>
+#include <range/v3/range_concepts.hpp>
+#include <range/v3/view/all.hpp>
+#include <range/v3/view/view.hpp>
+#include <range/v3/view_interface.hpp>
+#include <range/v3/view_adaptor.hpp>
+#include <range/v3/utility/functional.hpp>
+
+namespace ranges
+{
+    inline namespace v3
+    {
+        template<typename Rng, bool is_bidirectional = BidirectionalView<Rng>::value >
+        struct drop_last_view{};
+
+        template<typename Rng>
+        struct drop_last_view<Rng, true>
+          : view_interface<
+              drop_last_view<Rng, true>,
+              is_finite<Rng>::value ? finite : range_cardinality<Rng>::value   // finite at best
+          >
+        {
+            CONCEPT_ASSERT(BidirectionalView<Rng>());
+        private:
+            friend range_access;
+
+            using difference_t = range_difference_type_t<Rng>;
+            using size_t = range_size_type_t<Rng>;
+
+            Rng rng;
+            difference_t n;
+            detail::non_propagating_cache<sentinel_t<Rng>> sentinel;
+
+            template<class T>
+            size_t get_size(T&& rng) const
+            {
+                const size_t initial_size = ranges::size(rng);
+                const size_t n = static_cast<size_t>(this->n);
+
+                return initial_size > n
+                    ? initial_size - n
+                    : 0;
+            }
+        public:
+            drop_last_view() = default;
+            drop_last_view(Rng rng, difference_t n)
+              : rng(std::move(rng)), n(n)
+            {}
+
+            iterator_t<Rng> begin()
+            {
+                return ranges::begin(rng);
+            }
+            sentinel_t<Rng> end()
+            {
+                if (!sentinel)
+                {
+                    sentinel = ranges::prev(ranges::end(rng), n, ranges::begin(rng));
+                }
+                return *sentinel;
+            }
+            template<typename CRng = Rng const,
+                CONCEPT_REQUIRES_(RandomAccessRange<CRng>())>
+            iterator_t<CRng> begin() const
+            {
+                return ranges::begin(rng);
+            }
+            template<typename CRng = Rng const,
+                CONCEPT_REQUIRES_(RandomAccessRange<CRng>())>
+            sentinel_t<CRng> end() const
+            {
+                return ranges::prev(ranges::end(rng), n, ranges::begin(rng));
```

Comment:
This doesn't work if `ranges::end(rng)` returns a sentinel and not an iterator. You need to additionally require `CommonRange<CRng>()`. `begin() const` should too for consistency.

Reply:
Hmm if so... Shouldn't  non-const begin/end be CommonRange too? And since all begins/ends require CommonRange (both Forward and Bidi), maybe CommonRange should be required at view level?

BTW - `view::drop` need this "fix" too.